DM-38760: Initial version

rra opened the pull request:
Summarize current issues and some possible alternatives based on experimentation with Gafaelfawr.

rra committed:
Initial version

Summarize current issues and some possible alternatives based on
experimentation with Gafaelfawr.

jonathansick reviewed (approved):
These are great points, thanks.

In our analysis, I think we can treat the issues with the docstring format (numpydoc/napoleon) orthogonally from issues with API discovery (autodoc/automodapi). Here's how the responsibilities of the documentation stack break down, from my understanding:

1. Either napoleon or numpydoc preprocess the docstring into Sphinx roles
2. sphinx-autodoc-typehints also preprocesses the docstring to add type information to the docstring
3. automodapi's automodapi creates the stub pages for each API object and generates links to them with its own version of autosummary. Those stub pages contain [autodoc](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html) directives like `autoclass` or `autofunction` and within those `automethod`, etc. So when we're missing attributes, it's often the fault of automodapi. It's interesting how autodoc_pydantic works. Since it registers as an autodoc documenteer, automodapi automatically knows to use its directives for documenting a class. For example:

   ```rst
   GitHubBranchModel
   =================

   .. currentmodule:: safir.github.models

   .. autopydantic_model:: GitHubBranchModel
   ```

4. In turn, autodoc and autopydantic use the docstrings to put content on the page, typically using Sphinx directives like `class` for formatting

So I think any issues with getting types into the documentation can be linked to numpydoc/napoleon, but everything else is related to the automodapi/autodoc stack

For what it's worth, I recommended the Numpydoc format to the project because its a standard in scientific Python. I think it's been really useful and I don't support changing away from it given that the only substantial criticism is "having to repeat the return type." It could be interesting to try switching for specific circumstances, but I'm rather doubtful.

I think it's interesting how Sphinx's autosummary changes the documentation architecture. I'm a little skeptical of a whole module's APIs on one page. Does it use headings so that the in-page outline is useful? For example, a page like this one from aiohttp is really hard to use without an outline: https://docs.aiohttp.org/en/stable/web_reference.html

What about doing one `/api/*.rst` page per public module and using an `automodule` directive? at the top?

I'm a little surprised Sphinx's autosummary doesn't pick up `autodoc_pydantic` documenters; it seems like that could be fixed?

Overall I think the root of the problem is in improving/customizing how the documentation stub pages are created. Perhaps we could start committing the API stub pages and start maintaining them as necessary? You can use autosummary without generating new stub pages. You can even run the stub page generator separately, https://www.sphinx-doc.org/en/master/usage/extensions/autosummary.html#sphinx-autogen-generate-autodoc-stub-pages

Perhaps we could build on top of sphinx-autogen and customize it for our own purposes? Perhaps there's a way of marking parts of the API as being fully automatable, while other stubs we don't want auto-generated/overwritten by the tooling so that we can customize their behaviour?

rra commented:
Thank you for the detailed feedback! I've added your summary of how the current tools work in a new section, since that seemed like great information to capture. I also reorganized the known issues along the lines that you discuss, separating the ones that are related to API objects and page structure from the ones that are about how type information is extracted.

I don't have any answers to your questions (it looks like I didn't save the Gafaelfawr experimentation branch, alas, and I don't remember what the page structure was like), but I've also captured those in the tech note as potential avenues for future exploration.

I'm going to go ahead and merge this version based on your approval, but feel free to make additional changes, and also let me know if you want to be listed as a co-author; your contributions are far more than significant enough to warrant that!

rra committed:
Reorganize and flesh out from review

Reorganize the known issues to more cleanly separate by the involved
module. Move the idea to stop using the NumPy format to a rejected
alternatives section. Add an explanation of the current documentation
build process from Jonathan, and add open issues from the review.
